### ISSUE TITLE
Codecov incorrectly selects second git parent commit when Jenkins tests merged pull request

### ISSUE THREAD

**Issue by fdlk** [codecov/codecov-bash#137]
We run Jenkins CI on pull requests using the strategy "Merging the pull request with the current target branch revision" from the GitHub Branch Source Plugin:
<img width="737" alt="screen shot 2018-10-15 at 20 08 55" src="https://user-images.githubusercontent.com/1523838/46969480-762bb000-d0b6-11e8-8d12-4716f5fff37e.png">

The build log shows the merge details:
```
Merging remotes/origin/master commit 5438dfd8f70f3a525eaab95f5dc5caab599d0129 into PR head commit 808aa3ab5e3b389207a73c4aed07e07c481958b9
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 808aa3ab5e3b389207a73c4aed07e07c481958b9
 > git merge 5438dfd8f70f3a525eaab95f5dc5caab599d0129 # timeout=10
 > git rev-parse HEAD^{commit} # timeout=10
Merge succeeded, producing eb0c575b675620e5bf1666bbe39177bdd0c07aea
Checking out Revision eb0c575b675620e5bf1666bbe39177bdd0c07aea (PR-210)
```
So the PR head `808aa3a` gets checked out and the master HEAD `5438dfd` gets merged into it, producing merge commit `eb0c575`.

We then run the codecov script. It correctly figures out that we've tested a merge commit. But then it performs some magic to try and determine the SHA of the merge commit. This magic fails.

`git show --no-patch --format="%P"` returns `808aa3ab5e3b389207a73c4aed07e07c481958b9 5438dfd8f70f3a525eaab95f5dc5caab599d0129` and codecov incorrectly assumes that the second commit must be the commit ID under test. It sends the coverage information to the server with commit ID `5438dfd`, which is actually the SHA of the master HEAD.

This in turn causes codecov to decide that the report must belong to the last merged pull request, which messes up the PR coverage administration.

**Comment by fdlk** [comment]
I expect there is no easy way for codecov to figure out which is the correct parent commit and I'd prefer if it does less magic rather than incorrect magic.

**Comment by eps1lon** [comment]
CircleCI is affected by this as well: #134. At least CircleCI has the correct SHA already as an environment variable set. Maybe jenkins provides a similar helper?

**Comment by rolgalan** [comment]
We have exactly the same Jenkins configuration, and this problem keeps happening and it's killing our productivity. It fails most of the times in our builds.

I'm going to show a real example of my latest build, with my log of commits, in case this helps finding a solution.

We have base branch `develop`, which latest commit is the merge of `feature/ticket1`. From there we started two branches in parallel: ticket2 and ticket3. Ticket2 is merged successfully and then we open the Pull Request for Ticket3. This git log is produced from jenkins after the virtual merge:

```
*   9a3dc566c Merge commit '28f62c89d87b4c369859ecdd7cd27743a696a1d3' into HEAD
|\  
| *   28f62c89d Merge pull request #1003 from feature/ticket2 into develop
| |\  
| | * 38cd1c5f2 [ticket2] More code
| | * a2f673a33 [ticket2] Amazing feature
| | * 7331c658e [ticket2] Whatever
| |/  
* | 45c64e76d [ticket3] Set the branch explicitly in Codecov
* | f3db1e116 [ticket3] Log more info in the codecov stage to understand why it fails sometimes
|/  
*   180fc7c7e Merge pull request #997 from feature/ticket1 into develop
|\  
```

The magic to determine the head [lines 830 to 850](https://github.com/codecov/codecov-bash/blob/f107f3fc2229728d2db97ff087399bf331505f24/codecov#L830) correctly detects the merge, but the output of `mc=$(git show --no-patch --format="%P")` is `45c64e76d 28f62c89d`.

Then, **it always picks the second**, which in this case is develop HEAD. This is the problem. **It's not always the second**.  Today I've been trying merging locally into different branches, and even changing the time of my latest commit in the branch to be newer and older than the develop head, and I always get the first as the right one. I think sometimes the second would be the second would be the right one, but I don't know on what this depends.

We've added some logs in our CI before running the codecov script to try to understand when is the first or the second, because now I have no clue.

**Comment by thomasrockhu** [comment]
The merge commit SHA code has been removed except for GitHub Actions
